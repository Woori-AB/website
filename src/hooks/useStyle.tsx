import { useEffect } from "react";

export const useStyle = (keyframes: string) => {
  useEffect(() => {
    const style = document.createElement("style");
    style.innerHTML = keyframes;
    document.head.appendChild(style);
    return () => {
      document.head.removeChild(style);
    };
  }, [keyframes]);
};
