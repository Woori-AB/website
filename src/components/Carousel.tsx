"use client";

import Image from "next/image";
import { useStyle } from "@/hooks/useStyle";

export type Props = {
  images: string[];
  opaqueDuration: number;
  crossfadeDuration: number;
};

export default function Carousel({
  images,
  opaqueDuration,
  crossfadeDuration,
}: Props) {
  const frameDuration = opaqueDuration;
  const totalDuration = images.length * opaqueDuration;
  const fadeInEnd = crossfadeDuration / 2;
  const visibleEnd = fadeInEnd + frameDuration;
  const fadeOutEnd = visibleEnd + crossfadeDuration / 2;
  const fadeInEndPercent = (fadeInEnd / totalDuration) * 100;
  const visibleEndPercent = (visibleEnd / totalDuration) * 100;
  const fadeOutEndPercent = (fadeOutEnd / totalDuration) * 100;

  const keyframes = `
    @keyframes fade {
      0%   { opacity: 0; }
      ${fadeInEndPercent}%  { opacity: 1; }
      ${visibleEndPercent}% { opacity: 1; }
      ${fadeOutEndPercent}% { opacity: 0; }
      100% { opacity: 0; }
    }
  `;

  useStyle(keyframes);

  return (
    <div>
      {images.map((src, index) => (
        <div
          className="absolute w-full h-full opacity-0 ease-in-out fill-both"
          style={{
            animation: `fade ${totalDuration}s infinite`,
            animationDelay: `${index * opaqueDuration}s`,
          }}
          key={index}
        >
          <Image
            src={src}
            alt={`Slide ${index + 1}`}
            fill
            className="object-cover w-full h-full"
            sizes="100vw"
          />
        </div>
      ))}
    </div>
  );
}
