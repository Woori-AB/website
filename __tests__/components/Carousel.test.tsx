import Carousel from "@/components/Carousel";
import { cleanup, render } from "@testing-library/react";
import { describe, it, vi, expect, afterEach } from "vitest";
import * as useStyle from "@/hooks/useStyle";

describe("Carousel", () => {
  afterEach(() => {
    cleanup();
    vi.clearAllMocks();
  });

  it("matches snapshot(s)", () => {
    // Arrange
    const useStyleSpy = vi.spyOn(useStyle, "useStyle");

    // Act
    const { container } = render(
      <Carousel
        images={["/foo.jpg", "/bar.jpg"]}
        opaqueDuration={5}
        crossfadeDuration={3}
      />
    );

    // Assert
    expect(container).toMatchSnapshot();
    expect(useStyleSpy).toMatchSnapshot();
  });
});
