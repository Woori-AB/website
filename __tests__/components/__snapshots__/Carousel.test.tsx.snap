// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Carousel > matches snapshot(s) 1`] = `
<div>
  <div>
    <div
      class="absolute w-full h-full opacity-0 ease-in-out fill-both"
      style="animation: fade 10s infinite; animation-delay: 0s;"
    >
      <mocked-image
        alt="Slide 1"
        class="object-cover w-full h-full"
        fill=""
        sizes="100vw"
        src="/foo.jpg"
      />
    </div>
    <div
      class="absolute w-full h-full opacity-0 ease-in-out fill-both"
      style="animation: fade 10s infinite; animation-delay: 5s;"
    >
      <mocked-image
        alt="Slide 2"
        class="object-cover w-full h-full"
        fill=""
        sizes="100vw"
        src="/bar.jpg"
      />
    </div>
  </div>
</div>
`;

exports[`Carousel > matches snapshot(s) 2`] = `
[MockFunction] {
  "calls": [
    [
      "
    @keyframes fade {
      0%   { opacity: 0; }
      15%  { opacity: 1; }
      65% { opacity: 1; }
      80% { opacity: 0; }
      100% { opacity: 0; }
    }
  ",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
